// codemirror-lang-latex/dist/linter.js
import { syntaxTree } from '@codemirror/language';
// Performs basic syntax checking and best practices validation
export function latexLinter(options = {}) {
    const defaultOptions = {
        checkMissingDocumentEnv: true,
        checkUnmatchedEnvironments: true,
        checkMissingReferences: true
    };
    const opts = { ...defaultOptions, ...options };
    return (view) => {
        const diagnostics = [];
        const tree = syntaxTree(view.state);
        const doc = view.state.doc;
        // Check if document environment is missing
        if (opts.checkMissingDocumentEnv) {
            let hasDocumentEnv = false;
            tree.cursor().iterate(node => {
                if (node.name === 'DocumentEnvironment' || node.name === 'DocumentEnvName') {
                    hasDocumentEnv = true;
                    return false; // Stop iteration
                }
            });
            if (!hasDocumentEnv && doc.length > 100) { // Only apply to non-trivial documents
                diagnostics.push({
                    from: 0,
                    to: Math.min(doc.length, 200),
                    severity: 'warning',
                    message: 'Missing document environment. LaTeX documents should be enclosed in \\begin{document}...\\end{document}',
                    source: 'LaTeX'
                });
            }
        }
        // Check for unmatched environments
        if (opts.checkUnmatchedEnvironments) {
            const beginEnvs = new Map();
            const endEnvs = new Map();
            // Collect all begin and end environments
            tree.cursor().iterate(node => {
                var _a, _b;
                if (node.name === 'BeginEnv') {
                    const envNameNode = findEnvName(node);
                    if (envNameNode) {
                        const envName = doc.sliceString(envNameNode.from, envNameNode.to);
                        if (!beginEnvs.has(envName)) {
                            beginEnvs.set(envName, []);
                        }
                        (_a = beginEnvs.get(envName)) === null || _a === void 0 ? void 0 : _a.push(node.from);
                    }
                }
                else if (node.name === 'EndEnv') {
                    const envNameNode = findEnvName(node);
                    if (envNameNode) {
                        const envName = doc.sliceString(envNameNode.from, envNameNode.to);
                        if (!endEnvs.has(envName)) {
                            endEnvs.set(envName, []);
                        }
                        (_b = endEnvs.get(envName)) === null || _b === void 0 ? void 0 : _b.push(node.from);
                    }
                }
            });
            // Check for mismatches
            for (const [envName, begins] of beginEnvs.entries()) {
                const ends = endEnvs.get(envName) || [];
                if (begins.length > ends.length) {
                    // More begins than ends
                    for (let i = ends.length; i < begins.length; i++) {
                        diagnostics.push({
                            from: begins[i],
                            to: begins[i] + 6 + envName.length + 1,
                            severity: 'error',
                            message: `Missing \\end{${envName}}`,
                            source: 'LaTeX'
                        });
                    }
                }
            }
            for (const [envName, ends] of endEnvs.entries()) {
                const begins = beginEnvs.get(envName) || [];
                if (ends.length > begins.length) {
                    // More ends than begins
                    for (let i = begins.length; i < ends.length; i++) {
                        diagnostics.push({
                            from: ends[i],
                            to: ends[i] + 4 + envName.length + 1,
                            severity: 'error',
                            message: `Missing \\begin{${envName}}`,
                            source: 'LaTeX'
                        });
                    }
                }
            }
        }
        // Check for references to missing labels
        if (opts.checkMissingReferences) {
            const labels = new Set();
            const refs = new Map();
            // Collect all labels and refs
            tree.cursor().iterate(node => {
                var _a;
                if (node.name === 'LabelCtrlSeq') {
                    const labelArg = node.node.nextSibling;
                    if (labelArg && labelArg.name === 'LabelArgument') {
                        const content = doc.sliceString(labelArg.from + 1, labelArg.to - 1);
                        labels.add(content);
                    }
                }
                else if (node.name === 'RefCtrlSeq' || node.name === 'RefStarrableCtrlSeq') {
                    const refArg = (_a = node.node.nextSibling) === null || _a === void 0 ? void 0 : _a.nextSibling;
                    if (refArg && refArg.name === 'RefArgument') {
                        const content = doc.sliceString(refArg.from + 1, refArg.to - 1);
                        refs.set(content, node.from);
                    }
                }
            });
            // Check for refs to missing labels
            for (const [ref, pos] of refs.entries()) {
                if (!labels.has(ref)) {
                    diagnostics.push({
                        from: pos,
                        to: pos + ref.length + 6,
                        severity: 'warning',
                        message: `Reference to undefined label: ${ref}`,
                        source: 'LaTeX'
                    });
                }
            }
        }
        return diagnostics;
    };
}
// Helper function to find environment name node
function findEnvName(node) {
    // Look for the EnvNameGroup child
    let envNameGroup = null;
    for (let child = node.node.firstChild; child; child = child.nextSibling) {
        if (child.name === 'EnvNameGroup') {
            envNameGroup = child;
            break;
        }
    }
    if (envNameGroup) {
        for (let child = envNameGroup.firstChild; child; child = child.nextSibling) {
            if (child.name === 'EnvName' ||
                child.name === 'DocumentEnvName' ||
                child.name === 'TabularEnvName' ||
                child.name === 'EquationEnvName' ||
                child.name === 'EquationArrayEnvName' ||
                child.name === 'VerbatimEnvName' ||
                child.name === 'TikzPictureEnvName' ||
                child.name === 'FigureEnvName' ||
                child.name === 'ListEnvName' ||
                child.name === 'TableEnvName') {
                return child;
            }
        }
    }
    return null;
}
